### PR TITLE
Added support of macros

### PR DESCRIPTION
**Pull request by shaharyarn**

### CODE REVIEW

File: marine.c [hunk 1]
```diff
@@ -603,6 +629,11 @@ WS_DLL_PUBLIC int marine_add_filter(char *bpf, char *dfilter, char **fields, siz
         }
     }
 
+    macro_indices_copy = (int *) g_malloc0(sizeof(int) * fields_len);
+    for (size_t i = 0; i < fields_len; i++) {
```
Comment (yehonatanz) [inline marine.c]: `memcpy`?

File: marine_benchmark.c [hunk 2]
```diff
@@ -126,6 +127,8 @@ int main(int argc, char *argv[]) {
             "tcp.srcport",
             "udp.srcport"
     };
+    int three_no_macro[3];
+    memset(three_no_macro, 0, sizeof(three_no_macro));
```
Comment (yehonatanz) [inline marine_benchmark.c]: Pretty sure `int three_no_macro[3] = {0};` will work

File: marine.c [hunk 3]
```diff
@@ -285,12 +294,28 @@ marine_write_specified_fields(packet_filter *filter, epan_dissect_t *edt, char *
...
             if (fields->quote != '\0') {
                 output[counter++] = fields->quote;
             }
-            g_ptr_array_free(fv_p, TRUE);  /* get ready for the next packet */
+
+            i += filter->macro_indices[i];
+        }
+    }
+
+    /* get ready for the next packet */
```
Comment (yehonatanz) [inline marine.c]: Document that this loop is separate since the previoud loop might skip on fields due to macros

**Review by yehonatanz — CHANGES_REQUESTED** [review]
Had some cosmetic notes, but there's something more.
The `macro_indices` API is hard to understand, highly coupled to the implementation and does not make sense to an outside observer (API user). It took me a while to understand that the numbers represent "distance to the last field which is still part of the same macro".
Figure out a better way to tell the C code which fields are part of the same macro.
Some alternatives I thought about right now (can definitely be improved):
1. An array of booleans - `TRUE` if the field is the last in its group (or first, whatever) - will look like: `[0 0 1 0 1 1 1 0 0 0 0 1]`.
2. An array of ints that represent "group index" - will look like : `[0 0 0 0 1 2 2 3 4 5 5 5 5]`.

Either way, document your choice properly.

**Comment by shaharyarn** [comment]
Add an option to pass NULL to macro_indices to opt-out of the whole process (especially important if we are adding complexity by using different API).

_Resolved_

**Review by yehonatanz — CHANGES_REQUESTED** [review]
Looks good, but:
1. Can you provide some benchmark (probably on the python client is better), so we can estimate the performance cost of the new  , improved macros mechanism?
2. Consider renaming `macro_indices` to `macro_ids` and `last_macro` to `last_in_macro`.
3. Document that macros must be contiguous

**Comment by shaharyarn** [comment]
Before switching to c implementation:
before

After switching:
after

**Review by yehonatanz — APPROVED** [review]

File: marine_benchmark.c [hunk 4]
```diff
@@ -137,14 +138,15 @@ int main(int argc, char *argv[]) {
...
-            {"Benchmark with BPF, Display filter and eight extracted fields", bpf, dfilter, eight_fields, ARRAY_SIZE(eight_fields)},
+            {"Benchmark with BPF",                                            bpf,  NULL,    NULL,         NULL,           0},
+            {"Benchmark with Display filter",                                 NULL, dfilter, NULL,         NULL,           0},
+            {"Benchmark with BPF and Display filter",                         bpf,  dfilter, NULL,         NULL,           0},
+            {"Benchmark with three extracted fields",                         NULL, NULL,    three_fields, NULL, ARRAY_SIZE(three_fields)},
+            {"Benchmark with eight extracted fields",                         NULL, NULL,    eight_fields, NULL, ARRAY_SIZE(eight_fields)},
+            {"Benchmark with BPF, Display filter and three extracted fields", bpf,  dfilter, three_fields, NULL, ARRAY_SIZE(three_fields)},
+            {"Benchmark with BPF, Display filter and eight extracted fields", bpf,  dfilter, eight_fields, NULL, ARRAY_SIZE(eight_fields)},
```
Comment (tomlegkov) [inline marine_benchmark.c:148]: Very nice!
Optionally, I think you could add a benchmark case with macros (similar to the one you added in Python).

**Review by tomlegkov — APPROVED** [review]

**Review by yehonatanz — CHANGES_REQUESTED** [review]
Rebase your changes onto the current main branch to resolve your conflicts and fix your build.
Let's merge this :)

**Review by yehonatanz — APPROVED** [review]

**Pull request merged**